### PR TITLE
PSM-1496 Introduce `deploy-chart.sh` to deploy the `spinnaker` chart …

### PR DESCRIPTION
**Pull request by nathankooij**
…to Nexus

While there, introduce `.gitignore`.

### CODE REVIEW

File: deploy-chart.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+CHART_DIR="${CHARTS_DIR}/spinnaker"
+TMP_DIR="$(mktemp -d)"
+
+trap 'rm -rf -- "${TMP_DIR}"' INT TERM HUP EXIT
+helm3 package --dependency-update "${CHART_DIR}" --destination "${TMP_DIR}"
+find "${TMP_DIR}" -name '*.tgz' -exec \
+    curl --fail-with-body \
+    -u "${NEXUS_USERNAME}:${NEXUS_PASSWORD}" \
+    https://nexus.global.picnicinternational.com/repository/helm/ \
+    --upload-file {} +;
```
Comment (nathankooij) [inline deploy-chart.sh]: The `+` ensures errors from `find -exec` are propagated.

File: deploy-chart.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+NEXUS_PASSWORD="${NEXUS_PASSWORD:?NEXUS_PASSWORD not specified or empty}"
+CHARTS_DIR="${1:-charts}"
+
+CHART_DIR="${CHARTS_DIR}/spinnaker"
+TMP_DIR="$(mktemp -d)"
+
+trap 'rm -rf -- "${TMP_DIR}"' INT TERM HUP EXIT
+helm3 package --dependency-update "${CHART_DIR}" --destination "${TMP_DIR}"
+find "${TMP_DIR}" -name '*.tgz' -exec \
+    curl --fail-with-body \
```
Comment (nathankooij) [inline deploy-chart.sh]: `--fail-with-body` fails on any status code >= 400.

File: charts/spinnaker/Chart.yaml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 description: Open source, multi-cloud continuous delivery platform for releasing software changes with high velocity and confidence.
 name: spinnaker
-version: 2.2.12
+version: 2.2.12-picnic-1
```
Comment (nathankooij) [inline charts/spinnaker/Chart.yaml:4]: It's now already available on our Nexus. I can also diff with it against our internal deployment. (See JIRA for details.)

File: deploy-chart.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+NEXUS_USERNAME="${NEXUS_USERNAME:?NEXUS_USERNAME not specified or empty}"
+NEXUS_PASSWORD="${NEXUS_PASSWORD:?NEXUS_PASSWORD not specified or empty}"
+CHARTS_DIR="${1:-charts}"
+
+CHART_DIR="${CHARTS_DIR}/spinnaker"
+TMP_DIR="$(mktemp -d)"
+
+trap 'rm -rf -- "${TMP_DIR}"' INT TERM HUP EXIT
+helm3 package --dependency-update "${CHART_DIR}" --destination "${TMP_DIR}"
```
Comment (nathankooij) [inline deploy-chart.sh]: The `--dependency-update` flag ensures all dependencies from the `requirements.lock` are resolved.

**Comment by nathankooij** [comment]
Not sure if we want to follow our convention for commit messages here, seeing as this is a public fork. 🤔

**Review by ibabiankou — APPROVED** [review]
Looks quite straightforward 👍

File: deploy-chart.sh (hunk 2 above)
Comment (Badbond) [reply deploy-chart.sh]: Hmm. Ubuntu Focal (20.04) does ship an older `curl` that does not have this option :disappointed: The newer Ubuntu Jammy LTS (22.04) that came out a two months ago does. Any ideas for an alternative by any chance? :sweat_smile: If not, not a blocker as I don't have push permissions to Nexus anyway.

**Review by Badbond — COMMENTED** [review]
Functionality looking good. Regarding target branch, I believe convention is to have a `fork/<version>` branch (looking at [EP](https://github.com/PicnicSupermarket/error-prone/tree/fork/v2.14.0) & [NewRelic](https://github.com/PicnicSupermarket/newrelic-java-agent/branches)). Shall we do the same here?

**Review by Badbond — APPROVED** [review]
Thanks for picking this up and pushing the first version @nathankooij :pray:

File: deploy-chart.sh (hunk 2 above)
Comment (nathankooij) [reply deploy-chart.sh]: I pushed a few changes for this. Can you try it please? :)

**Review by Badbond — APPROVED** [review]
Yes! Looking good! Received a 403 as expected.

```
> PUT /repository/helm/spinnaker-2.2.12-picnic-1.tgz HTTP/2
> Host: nexus.global.picnicinternational.com
> authorization: Basic <not-for-you>
> user-agent: curl/7.68.0
> accept: */*
> content-length: 126091
> 
{ [5 bytes data]
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
} [5 bytes data]
* We are completely uploaded and fine
{ [5 bytes data]
< HTTP/2 403 
< date: Mon, 11 Jul 2022 17:17:28 GMT
< server: Nexus/3.38.1-01 (OSS)
< x-content-type-options: nosniff
< content-security-policy: sandbox allow-forms allow-modals allow-popups allow-presentation allow-scripts allow-top-navigation
< x-xss-protection: 1; mode=block
< 
{ [5 bytes data]
100  123k    0     0  100  123k      0   549k --:--:-- --:--:--
 --:--:--  549k
* Connection #0 to host nexus.global.picnicinternational.com left intact
Failed to publish
```

**Comment by nathankooij** [comment]
> Yes! Looking good! Received a 403 as expected.
> 
> ```
> > PUT /repository/helm/spinnaker-2.2.12-picnic-1.tgz HTTP/2
> > Host: nexus.global.picnicinternational.com
> > authorization: Basic <not-for-you>
> > user-agent: curl/7.68.0
> > accept: */*
> > content-length: 126091
> > 
> { [5 bytes data]
> * Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
> } [5 bytes data]
> * We are completely uploaded and fine
> { [5 bytes data]
> < HTTP/2 403 
> < date: Mon, 11 Jul 2022 17:17:28 GMT
> < server: Nexus/3.38.1-01 (OSS)
> < x-content-type-options: nosniff
> < content-security-policy: sandbox allow-forms allow-modals allow-popups allow-presentation allow-scripts allow-top-navigation
> < x-xss-protection: 1; mode=block
> < 
> { [5 bytes data]
> 100  123k    0     0  100  123k      0   549k --:--:-- --:--:--
>  --:--:--  549k
> * Connection #0 to host nexus.global.picnicinternational.com left intact
> Failed to publish
> ```

Nice! Thanks for checking.

**Pull request merged**